### PR TITLE
More explicit build instructions

### PR DESCRIPTION
**Pull request by RalfJung**
@ssomers does this help, or do you think there is more we should say?

Fixes https://github.com/rust-lang/miri/issues/1306

### CODE REVIEW

**Comment by ssomers** [comment]
I still don't like the two titles "Building Miri with". They are more like "Configuring the working tree for". Here's how I would arrange things:

## Getting started
## _(new header)_ Preparing the working tree _or_ Configuring the build settings
### ~~Building Miri~~ with a pre-built rustc
### _(moved up)_ ~~Building Miri~~ with a locally built rustc
## Building and testing
_(the paragraph you added)_
### Fixing Miri when rustc changes
### Testing the Miri driver
### Testing `cargo miri`

**Comment by RalfJung** [comment]
Hm... I see. That makes sense.

The reason I structured things the way they are is that "with a locally built rustc" is a rare thing to do (I think), so I wanted it to be last in the docs. Your proposal as you said moves it up, putting it "in the way" of what the normal reader will need (which is "pre-build rustc" followed by "building an testing"). What do you think about that?

**Comment by ssomers** [comment]
You're right, I guess most people don't start off with a freshly built rust working tree, and I noticed myself it's not a good idea to ignore the rust-version file and "just compile with what I have". And it's good to have a quick ramp up to a working miri. But then I don't understand why "Fixing Miri when rustc changes" comes so early, spoiling the fun before it begins.


> 
> Hm... I see. That makes sense.
> 
> The reason I structured things the way they are is that "with a locally built rustc" is a rare thing to do (I think), so I wanted it to be last in the docs. Your proposal as you said moves it up, putting it "in the way" of what the normal reader will need (which is "pre-build rustc" followed by "building an testing"). What do you think about that?

**Comment by RalfJung** [comment]
> But then I don't understand why "Fixing Miri when rustc changes" comes so early, spoiling the fun before it begins.

I think I did that because I thought that might be a good on-ramping exercise for new contributors. But I do not think that any more, so I agree this should be rearranged.

**Comment by RalfJung** [comment]
> I noticed myself it's not a good idea to ignore the rust-version file and "just compile with what I have"

Indeed, and somehow I hope to write the docs in a way to convey this most clearly. :)

**Comment by RalfJung** [comment]
@ssomers I did some more editing, please let me know what you think.

@oli-obk I'd also be interested in your feedback, I hope you are okay with these edits.

**Comment by oli-obk** [comment]
Yes, it makes a lot of sense to move the rustup instructions to later, this isn't what ppl will be looking for the most.

**Comment by RalfJung** [comment]
However, I am not sure if they really make sense in the "Building and testing Miri" section... they are conceptually more like setting up the build environment.

**Comment by oli-obk** [comment]
Setting up the build environment does not require users to know anything about the correct nightly. We have a toolchain file that will automatically give them the correct nightly if they use rustup. Only if you want to modify that file, you need to know about the other steps. Maybe it should just be its own section?

**Comment by RalfJung** [comment]
So what about this?

> We have a toolchain file that will automatically give them the correct nightly if they use rustup.

Not exactly (as we pin by commit, not by nightly), but we have `rustup-toolchain` to make that happen.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -25,42 +25,41 @@ install that exact version of rustc as a toolchain:
...
 
-## Testing the Miri driver
-[testing-miri]: #testing-the-miri-driver
+### Testing the Miri driver
 
-The Miri driver in the `miri` binary is the "heart" of Miri: it is basically a
-version of `rustc` that, instead of compiling your code, runs it.  It accepts
-all the same flags as `rustc` (though the ones only affecting code generation
-and linking obviously will have no effect) [and more][miri-flags].
+The Miri driver in the `src/bin/miri.rs` binary is the "heart" of Miri: it is
```
Comment (ssomers) [inline CONTRIBUTING.md:48]: "the `src/bin/miri.rs` binary" sounds strange to me - is it the binary made from ...?

Comment (RalfJung) [reply CONTRIBUTING.md]: I adjusted the wording, better like this?

> The Miri driver compiled from `src/bin/miri.rs` is the "heart" of Miri:

**Pull request merged**